xrootd/protocol/marshaling.go

```diff
@@ -11,6 +11,32 @@ import (
 	"github.com/pkg/errors"
 )
 
+// Request is a XRootD request issued to a server.
+type Request interface {
+	// ReqID uniquely identifies the type of a request.
+	ReqID() uint16
+	Marshaler
+	Unmarshaler
+}
+
+// Marshaler is the interface implemented by a type that can marshal itself
+// into a binary form, following the XRootD protocol.
+//
+// MarshalXrd encodes the receiver into a binary form and returns the result.
+type Marshaler interface {
+	MarshalXrd() ([]byte, error)
+}
+
+// Unmarshaler is the interface implemented by a type that can
+// unmarshal a binary representation of itself, following the XRootD protocol.
+//
+// UnmarshalXrd must be able to decode the form generated by MarshalXrd.
+// UnmarshalXrd must copy the data if it wishes to retain the data after
+// returning.
+type Unmarshaler interface {
+	UnmarshalXrd(data []byte) (int, error)
```

Comment:
ok... to address this annoying (to me, at least) issue of this extra `int` return value, what about modifying the (un)marshalers to something like:
```go
type Marshaler interface {
    MarshalXrd(buf *Buffer) error
}

type Unmarshaler interface {
    UnmarshalXrd(buf *Buffer) error
}
```

where `Buffer` is the union of the current `Encoder` and `Decoder` type.

that way, the current issue of unmarshaling `protocol/protocol.Request`:
```go
// UnmarshalXrd implements xrootd/protocol.Unmarshaler
func (o *Response) UnmarshalXrd(data []byte) (pos int, err error) {
	dec := xrdenc.NewDecoder(data)
	o.BinaryProtocolVersion = dec.ReadI32()
	o.flags = flags(dec.ReadI32())
	if dec.Available() > 0 {
		dec.Skip(1)
		dec.Skip(1)
		o.SecurityVersion = dec.ReadU8()
		o.SecurityOptions = securityOptions(dec.ReadU8())
		o.SecurityLevel = protocol.SecurityLevel(dec.ReadU8())
		o.SecurityOverrides = make([]protocol.SecurityOverride, dec.ReadU8())
		for i := 0; i < len(o.SecurityOverrides); i++ {
			n, err := o.SecurityOverrides[i].UnmarshalXrd(dec.Bytes())
			if err != nil {
				return 0, err
			}
			dec.Skip(n)
		}
	}
	return dec.Pos(), err
}
```
would become:
```go
// UnmarshalXrd implements xrootd/protocol.Unmarshaler
func (o *Response) UnmarshalXrd(buf *xrdproto.Buffer) error {
	o.BinaryProtocolVersion = buf.ReadI32()
	o.flags = flags(buf.ReadI32())
	if buf.Len() > 0 {
		buf.Skip(1)
		buf.Skip(1)
		o.SecurityVersion = buf.ReadU8()
		o.SecurityOptions = securityOptions(buf.ReadU8())
		o.SecurityLevel = protocol.SecurityLevel(buf.ReadU8())
		o.SecurityOverrides = make([]protocol.SecurityOverride, buf.ReadU8())
		for i := 0; i < len(o.SecurityOverrides); i++ {
			err := o.SecurityOverrides[i].UnmarshalXrd(buf)
			if err != nil {
				return err
			}
		}
	}
	return nil
}
```

most notably, the `dec.Skip(n)` statement (which could easily be overlooked and fogotten) just disappeared.

what do you think?

Reply:
What do you think about keeping `Encoder` and `Decoder` and passing them instead of `xrdproto.Buffer`? Because we don't need to perform any writing in `UnmarshalXrd` and any reading in `MarshalXrd`.
So it will look like:
```go
type Marshaler interface {
	MarshalXrd(enc *xrdenc.Encoder) error
}

type Unmarshaler interface {
	UnmarshalXrd(dec *xrdenc.Decoder) error
}
```

Reply:
Then, what about renaming 'Decoder' into 'RBuffer' and 'Encoder' into 'WBuffer'?

Reply:
Then, what about renaming 'Decoder' into 'RBuffer' and 'Encoder' into 'WBuffer'?

Reply:
Sounds good!